kernel/work.c

```diff
@@ -0,0 +1,966 @@
+	 *
+	 * If not successful there is no notification that the work has been
+	 * abandoned.  Sorry.
+	 */
+	if (atomic_test_and_clear_bit(&wp->flags, K_WORK_DELAYED_BIT)) {
+		queue = dw->queue;
+		(void)submit_to_queue_locked(wp, &queue);
+	}
+
+	k_spin_unlock(&lock, key);
+}
+
+void k_work_init_delayable(struct k_work_delayable *dwork,
+			    k_work_handler_t handler)
+{
+	__ASSERT_NO_MSG(dwork != NULL);
+	__ASSERT_NO_MSG(handler != 0);
+
+	*dwork = (struct k_work_delayable){
+		.work = {
+			.handler = handler,
+			.flags = K_WORK_DELAYABLE,
+		},
+	};
+	z_init_timeout(&dwork->timeout);
+	(void)work_timeout;
+}
+
+/* Attempt to schedule a work item for future (maybe immediate)
+ * submission.
+ *
+ * Invoked with work lock held.
+ *
+ * See also submit_to_queue_locked(), which implements this for a no-wait
+ * delay.
+ *
+ * Invoked with work lock held.
+ *
+ * @param queuep pointer to a pointer to a queue.  On input this
+ * should dereference to the proposed queue (which may be null); after
+ * completion it will be null if the work was not submitted or if
+ * submitted will reference the queue it was submitted to.  That may
+ * or may not be the queue provided on input.
+ *
+ * @param dwork the delayed work structure
+ *
+ * @param delay the delay to use before scheduling.
+ *
+ * @retval from submit_to_queue_locked() if delay is K_NO_WAIT; otherwise
+ * @retval 1 to indicate successfully scheduled.
+ */
+int schedule_to_queue_locked(struct k_work_q **queuep,
+			     struct k_work_delayable *dwork,
+			     k_timeout_t delay)
+{
+	int ret = 1;
+	struct k_work *work = &dwork->work;
+
+	if (K_TIMEOUT_EQ(delay, K_NO_WAIT)) {
+		return submit_to_queue_locked(work, queuep);
+	}
+
+	atomic_set_bit(&work->flags, K_WORK_DELAYED_BIT);
+	dwork->queue = *queuep;
+
+	/* Add timeout */
+	z_add_timeout(&dwork->timeout, work_timeout, delay);
+
+	return ret;
+}
+
+/* Unschedule delayable work.
+ *
+ * If the work is delayed, cancel the timeout and clear the delayed
+ * flag.
+ *
+ * Invoked with work lock held.
+ *
+ * @param dwork pointer to delayable work structure.
+ *
+ * @return true iff work had been delayed
+ */
+static inline bool unschedule_locked(struct k_work_delayable *dwork)
+{
+	bool ret = false;
+	struct k_work *work = &dwork->work;
+
+	/* If scheduled, try to cancel. */
+	if (atomic_test_and_clear_bit(&work->flags, K_WORK_DELAYED_BIT)) {
+		z_abort_timeout(&dwork->timeout);
+		ret = true;
+	}
+
+	return ret;
+}
+
+/* Full cancellation of a delayable work item.
+ *
+ * Unschedules the delayed part then delegates to standard work
+ * cancellation.
+ *
+ * Invoked with work lock held.
+ *
+ * @param dwork delayable work item
+ *
+ * @return k_work_busy_get() flags
+ */
+static int cancel_delayable_async_locked(struct k_work_delayable *dwork)
+{
+	(void)unschedule_locked(dwork);
+
+	return cancel_async_locked(&dwork->work);
+}
+
+int k_work_schedule_for_queue(struct k_work_q *queue,
+			       struct k_work_delayable *dwork,
```

Comment:
Yeah.  There is/was a subtle difference at one point but rather than explain it it's easier to rename the internal function to match.  (And make it `static`, which it somehow wasn't.)